sklearnex/tests/_utils.py

```diff
@@ -0,0 +1,148 @@
+
+from inspect import isclass
+
+import numpy as np
+from sklearn.base import (
+    BaseEstimator,
+    ClassifierMixin,
+    ClusterMixin,
+    OutlierMixin,
+    RegressorMixin,
+    TransformerMixin,
+)
+from sklearn.datasets import load_diabetes, load_iris
+from sklearn.neighbors._base import KNeighborsMixin
+
+from onedal.tests.utils._dataframes_support import _convert_to_dataframe
+from sklearnex import get_patch_map, patch_sklearn, unpatch_sklearn
+from sklearnex.neighbors import (
+    KNeighborsClassifier,
+    KNeighborsRegressor,
+    LocalOutlierFactor,
+    NearestNeighbors,
+)
+from sklearnex.svm import SVC, NuSVC
+
+
+def _load_all_models(patched, estimator=True):
+
+    try:
+        if patched:
+            patch_sklearn()
+
+        models = {}
+        for patch_infos in get_patch_map().values():
+            candidate = getattr(patch_infos[0][0][0], patch_infos[0][0][1])
+            if candidate is not None and isclass(candidate) == estimator:
+                if not estimator or issubclass(candidate, BaseEstimator):
+                    models[patch_infos[0][0][1]] = candidate
+    finally:
+        if patched:
+            unpatch_sklearn()
+
+    return models
+
+
+PATCHED_MODELS = _load_all_models(patched=True)
+UNPATCHED_MODELS = _load_all_models(patched=False)
+
+PATCHED_FUNCTIONS = _load_all_models(patched=True, estimator=False)
+UNPATCHED_FUNCTIONS = _load_all_models(patched=True, estimator=False)
+
+mixin_map = [
+    [
+        ClassifierMixin,
+        ["decision_function", "predict", "predict_proba", "predict_log_proba", "score"],
+        "classification",
+    ],
+    [RegressorMixin, ["predict", "score"], "regression"],
+    [ClusterMixin, ["fit_predict"], "classification"],
+    [TransformerMixin, ["fit_transform", "transform", "score"], "classification"],
+    [OutlierMixin, ["fit_predict", "predict"], "classification"],
+    [KNeighborsMixin, ["kneighbors"], None],
+]
+
+
+SPECIAL_INSTANCES = {
+    i.__str__(): i
+    for i in [
+        LocalOutlierFactor(novelty=True),
+        SVC(probability=True),
+        NuSVC(probability=True),
+        KNeighborsClassifier(algorithm="brute"),
+        KNeighborsRegressor(algorithm="brute"),
+        NearestNeighbors(algorithm="brute"),
+    ]
+}
+
+
+def gen_models_info(algorithms):
+    output = []
+    for i in algorithms:
+        # split handles SPECIAL_INSTANCES or custom inputs
+        # custom sklearn inputs must be a dict of estimators
+        # with keys set by the __str__ method
+        est = UNPATCHED_MODELS[i.split("(")[0]]
+
+        methods = []
+        candidates = set(
+            [i for i in dir(est) if not i.startswith("_") and not i.endswith("_")]
+        )
+
+        for mixin, method, _ in mixin_map:
+            if issubclass(est, mixin):
+                methods += list(candidates.intersection(set(method)))
+
+        methods = list(set(methods))  # return only unique values
```

Comment:
This part looks overly complicated with all the conversion to sets and list comprehensions. Doesn't the below suggestion work as well? Can probably be simplified even more...

```suggestion
        methods = set()
        candidates = set(
            [i for i in dir(est) if not i.startswith("_") and not i.endswith("_")]
        )

        for mixin, method, _ in mixin_map:
            if issubclass(est, mixin):
                methods |= candidates & set(method)
```